### PR TITLE
[SPARK-19668][ML] Multiple NGram sizes

### PR DESCRIPTION
**Pull request by mpetruska**
## What changes were proposed in this pull request?

[Jira ticket](https://issues.apache.org/jira/browse/SPARK-19668):
- implements extraction of multiple sizes of ngrams with `feature.NGram`

## How was this patch tested?

- unit tests were added for the implementation (`multiSliding` function)
- test cases were added to `NGramSuite` and `Word2VecSuite`

----

Benchmark of the implementation of the `multiSliding` function against a simpler, but naive implementation (tested on 1 machine):

```Scala
scala> def multiSlidingRepeated[A](x: Seq[A], min: Int, max: Int): Seq[Seq[A]] = (min to max).flatMap(x.sliding)
multiSlidingRepeated: [A](x: Seq[A], min: Int, max: Int)Seq[Seq[A]]

scala> time(multiSliding(1 to 10000, 5, 20)).toMillis
res5: Long = 66

scala> time(multiSlidingRepeated(1 to 10000, 5, 20)).toMillis
res6: Long = 434

scala> time(multiSliding(1 to 50000, 5, 20)).toMillis
res7: Long = 595

scala> time(multiSlidingRepeated(1 to 50000, 5, 20)).toMillis
java.lang.OutOfMemoryError: GC overhead limit exceeded
  ...

scala> time(multiSliding(1 to 10000, 5, 30)).toMillis
res9: Long = 91

scala> time(multiSlidingRepeated(1 to 10000, 5, 30)).toMillis
res11: Long = 428
```

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/ml/extensions/seq/package.scala [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+   *   Seq(4, 5), Seq(4, 5, 6), Seq(4, 5, 6, 7), Seq(4, 5, 6, 7, 8),
+   *   Seq(5, 6), Seq(5, 6, 7), Seq(5, 6, 7, 8), Seq(5, 6, 7, 8, 9),
+   *   Seq(6, 7), Seq(6, 7, 8), Seq(6, 7, 8, 9), Seq(6, 7, 8, 9, 10),
+   *   Seq(7, 8), Seq(7, 8, 9), Seq(7, 8, 9, 10),
+   *   Seq(8, 9), Seq(8, 9, 10),
+   *   Seq(9, 10)
+   * )
+   * }}}
+   */
+  @Since("From which version?")
```
Comment (mpetruska) [inline mllib/src/main/scala/org/apache/spark/ml/extensions/seq/package.scala]: Please help me fill this in.

File: mllib/src/main/scala/org/apache/spark/ml/extensions/seq/package.scala [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+      addWindowsFromFinalBuffer(accumulated, finalBuffer).reverse
+    }
+
+    (1 <= min) && (min <= max) match {
+      case true => calculateMultiSliding()
+      case false => Seq.empty
+    }
+  }
+
+  @Since("From which version?")
```
Comment (mpetruska) [inline mllib/src/main/scala/org/apache/spark/ml/extensions/seq/package.scala]: Note to self: please don't forget to fill this in.

File: mllib/src/main/scala/org/apache/spark/ml/extensions/seq/package.scala [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+  implicit class SeqOps[A](val x: Seq[A]) extends AnyVal {
+
+    /**
+     * Calculates sliding windows over multiple length parameters
+     * simultaneously.
+     * @see `org.apache.spark.ml.extensions.seq.multiSliding`
+     */
+    @Since("From which version?")
```
Comment (mpetruska) [inline mllib/src/main/scala/org/apache/spark/ml/extensions/seq/package.scala]: Note to self: please don't forget to fill this in.

File: mllib/src/main/scala/org/apache/spark/ml/feature/NGram.scala [hunk 4]
```diff
@@ -42,11 +42,22 @@ class NGram @Since("1.5.0") (@Since("1.5.0") override val uid: String)
...
+  val n: IntParam = new IntParam(this, "n", "minimum number of elements per n-gram (>=1)",
+    ParamValidators.gtEq(1))
+
+  /**
+   * Maximum n-gram length, greater than or equal to `n`.
+   * All values of m such that n <= m <= maxN will be used.
+   * Default: 2, bigram features
+   * @group param
+   */
+  @Since("From which version?")
```
Comment (mpetruska) [inline mllib/src/main/scala/org/apache/spark/ml/feature/NGram.scala]: Note to self: please fill this in.

File: mllib/src/main/scala/org/apache/spark/ml/feature/NGram.scala [hunk 5]
```diff
@@ -57,10 +68,22 @@ class NGram @Since("1.5.0") (@Since("1.5.0") override val uid: String)
   @Since("1.5.0")
   def getN: Int = $(n)
 
+  /** @group setParam */
+  @Since("From which version?")
```
Comment (mpetruska) [inline mllib/src/main/scala/org/apache/spark/ml/feature/NGram.scala]: Note to self: please fill this in.

File: mllib/src/main/scala/org/apache/spark/ml/feature/NGram.scala [hunk 6]
```diff
@@ -57,10 +68,22 @@ class NGram @Since("1.5.0") (@Since("1.5.0") override val uid: String)
...
+  def setMaxN(value: Int): this.type = set(maxN, value)
+
+  /** @group getParam */
+  @Since("From which version?")
```
Comment (mpetruska) [inline mllib/src/main/scala/org/apache/spark/ml/feature/NGram.scala]: Note to self: please fill this in.

**Comment by holdenk** [comment]
Thanks for working on this. Jenkins OK to test.

**Comment by holdenk** [comment]
Let me CC @MLnick as well.

File: mllib/src/main/scala/org/apache/spark/ml/extensions/seq/package.scala (hunk 1 above)
Comment (holdenk) [inline mllib/src/main/scala/org/apache/spark/ml/extensions/seq/package.scala]: We probably don't want these APIs to be public unless there is a good reason to have them be. Maybe make them private?

File: mllib/src/main/scala/org/apache/spark/ml/feature/NGram.scala [hunk 7]
```diff
@@ -42,11 +42,22 @@ class NGram @Since("1.5.0") (@Since("1.5.0") override val uid: String)
...
+    ParamValidators.gtEq(1))
+
+  /**
+   * Maximum n-gram length, greater than or equal to `n`.
+   * All values of m such that n <= m <= maxN will be used.
+   * Default: 2, bigram features
+   * @group param
+   */
+  @Since("From which version?")
+  val maxN: IntParam = new IntParam(this, "maxN", "maximum number elements per n-gram (>=n)",
```
Comment (holdenk) [inline mllib/src/main/scala/org/apache/spark/ml/feature/NGram.scala]: Should this have a default value? Also perhaps the explanation that if unset maxN = n

File: mllib/src/test/scala/org/apache/spark/ml/feature/NGramSuite.scala [hunk 8]
```diff
@@ -76,11 +76,32 @@ class NGramSuite extends SparkFunSuite with MLlibTestSparkContext with DefaultRe
...
+    testNGram(nGram, dataset)
+  }
+
   test("read/write") {
     val t = new NGram()
       .setInputCol("myInputCol")
       .setOutputCol("myOutputCol")
       .setN(3)
+      .setMaxN(5)
     testDefaultReadWrite(t)
```
Comment (holdenk) [inline mllib/src/test/scala/org/apache/spark/ml/feature/NGramSuite.scala]: Would be good to make sure read/write makes sense even when maxN isn't set perhaps?

File: examples/src/main/python/ml/n_gram_example.py [hunk 9]
```diff
@@ -35,7 +35,7 @@
         (2, ["Logistic", "regression", "models", "are", "neat"])
     ], ["id", "words"])
 
-    ngram = NGram(n=2, inputCol="words", outputCol="ngrams")
+    ngram = NGram(n=2, maxN=4, inputCol="words", outputCol="ngrams")
```
Comment (holdenk) [inline examples/src/main/python/ml/n_gram_example.py]: This generally isn't enough to expose the new functionality in Python. Take a look in `feature.py` & also consider adding some Python tests.

File: mllib/src/main/scala/org/apache/spark/ml/extensions/seq/package.scala (hunk 1 above)
Comment (holdenk) [reply mllib/src/main/scala/org/apache/spark/ml/extensions/seq/package.scala]: Also is there a reason not to put this functionality inside of ngram its self for now?

File: mllib/src/main/scala/org/apache/spark/ml/extensions/seq/package.scala [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+
+    @tailrec
+    def addWindowsFromFinalBuffer(acc: List[B], buffer: Queue[A]): List[B] = {
+      buffer.dequeueOption match {
+        case Some((_, tail)) => addWindowsFromFinalBuffer(addWindowsFromBuffer(acc, tail), tail)
+        case None => acc
+      }
+    }
+
+    def calculateMultiSliding(): List[B] = {
```
Comment (holdenk) [inline mllib/src/main/scala/org/apache/spark/ml/extensions/seq/package.scala]: Does this perform as well as the current code does?

**Review by holdenk — COMMENTED** [review]
Thanks for working on this & your first PR to Spark  :) I've got some initial questions and comments, but I haven't dug too deeply. I'm sure @MLnick will have more thoughts as well :)

File: mllib/src/main/scala/org/apache/spark/ml/extensions/seq/package.scala (hunk 1 above)
Comment (mpetruska) [reply mllib/src/main/scala/org/apache/spark/ml/extensions/seq/package.scala]: Right, agreed, will move the implementation into `NGram`.

File: mllib/src/main/scala/org/apache/spark/ml/feature/NGram.scala (hunk 7 above)
Comment (mpetruska) [reply mllib/src/main/scala/org/apache/spark/ml/feature/NGram.scala]: Added explanation in the scaladoc comment for the case when `maxN` is not set.
The default is set to 2 here: https://github.com/apache/spark/pull/19659

File: mllib/src/main/scala/org/apache/spark/ml/extensions/seq/package.scala (hunk 10 above)
Comment (mpetruska) [reply mllib/src/main/scala/org/apache/spark/ml/extensions/seq/package.scala]: Nope, unfortunately it's not the case:

```Scala
scala> def slidingOriginal(x: Seq[String], n: Int): Seq[String] = {
     |     x.iterator.sliding(n).withPartial(false).map(_.mkString(" ")).toSeq
     |   }
slidingOriginal: (x: Seq[String], n: Int)Seq[String]

scala> def slidingNew(x: Seq[String], n: Int): Seq[String] = {
     |     multiSliding(x, n, n).map(_.mkString(" "))
     |   }
slidingNew: (x: Seq[String], n: Int)Seq[String]

scala> time(slidingOriginal((1 to 10000).map(_.toString), 5))
res9: scala.concurrent.duration.FiniteDuration = 2427473 nanoseconds

scala> time(slidingOriginal((1 to 10000).map(_.toString), 5)).toMillis
res10: Long = 3

scala> time(slidingNew((1 to 10000).map(_.toString), 5)).toMillis
res11: Long = 15

scala> time(slidingOriginal((1 to 100000).map(_.toString), 5)).toMillis
res12: Long = 11

scala> time(slidingNew((1 to 100000).map(_.toString), 5)).toMillis
res13: Long = 479

scala> time(slidingOriginal((1 to 10000).map(_.toString), 50)).toMillis
res14: Long = 1

scala> time(slidingNew((1 to 10000).map(_.toString), 50)).toMillis
res15: Long = 136
```

Adding code to fall back to the original implementation if `n == maxN`...

**Comment by mpetruska** [comment]
@holdenk : Thanks for the review.
Fixed/addressed your comments, can you please have another look?
Also I'd like some help with regards to the `@Since` attributes; how should I fill them in? Should I fill them in at all?

**Comment by mpetruska** [comment]
Also, will this work in SparkR, or do I need to add some code/configuration to expose the functionality to [R]?

**Comment by holdenk** [comment]
So SparkR will have to be separate, I'm going to cc @felixcheung , I don't know if we currently expose this functionality in the Spark R side.

As for the @Since attributes you choose them based on the version that your code is going to first appear in. At the point it might be 2.3.0 but it might also not. I'll try and do a more thorough review next week.

**Comment by holdenk** [comment]
Is this a thing you are still working on @mpetruska ?

**Comment by mpetruska** [comment]
No, I'm not working on this one, the feature seems to lack any real interest. We can close it, if you like.

**Comment by mgaido91** [comment]
@holdenk I can take this up if this is needed. Let me know. Thanks.

**Comment by holdenk** [comment]
I haven't seen much demand for it, so I think closing it is probably reasonable. We can re-open the issue if there is demand.

**Comment by holdenk** [comment]
Although if you have a use case/ are interested in this @mgaido91  I'm happy to help with the review.

**Comment by mgaido91** [comment]
I think we can close it and reopen if someone express new interest on this, thanks @holdenk

**Pull request closed without merging**